### DISCUSSION TITLE
Opencv cannot run as usual

### DISCUSSION THREAD

**Discussion by Buruxianian** [zauberzeug/nicegui#2321] [Q&A] [answered]
### Question

Unable to obtain frame by running nicegui-main\examples\opencv_webcam\main.py

To solve the problem, I compared two simplified scripts. But I don't know how to solve it.
The first script was unable to retrieve frames, as detailed below：
```py
import base64
import cv2
from nicegui import ui

video_capture = cv2.VideoCapture(0)
ui_interactive_image = ui.interactive_image()

def lazy_update() -> None:
    global video_capture, ui_interactive_image
    ret, frame = video_capture.read()
    if ret and ui_interactive_image is not None:
        _, imencode_image = cv2.imencode('.jpg', frame)
        base64_image = base64.b64encode(imencode_image)
        base64_image_string = 'data:image/jpg;base64,' + base64_image.decode('ascii')
        ui_interactive_image.source = base64_image_string
        
ui.timer(interval=2, callback=lazy_update)
ui.run()
```

The second script can retrieve frames, as detailed below：
```py
import base64
import cv2
from nicegui import ui

ui_interactive_image = ui.interactive_image()

def lazy_update() -> None:
    global ui_interactive_image
    video_capture = cv2.VideoCapture(0)

    ret, frame = video_capture.read()
    if ret and ui_interactive_image is not None:
        _, imencode_image = cv2.imencode('.jpg', frame)
        base64_image = base64.b64encode(imencode_image)
        base64_image_string = 'data:image/jpg;base64,' + base64_image.decode('ascii')
        ui_interactive_image.source = base64_image_string
    video_capture.release()
ui.timer(interval=2, callback=lazy_update)

ui.run()
```

**Comment by rodja** [comment]
Have you seen and tried our [OpenCV webcam example](https://github.com/zauberzeug/nicegui/blob/main/examples/opencv_webcam/main.py)?

**Reply by Buruxianian** [reply]
I have tested it and cannot obtain frames.
It's strange that I can run the code below normally, but I can't use run opencv with nicegui.
```py
import cv2
cap = cv2.VideoCapture(0)
if not cap.isOpened():
    print("error")
else:
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        
        
        cv2.imshow("Frame", frame)
        if cv2.waitKey(1) == ord('q'):
            break
    
    cap.release()
    cv2.destroyAllWindows()
```

**Comment by Buruxianian** [comment, marked as answer]
The problem has been resolved.
```
ui.run(reload=False)
```
After turning off the reload, it will run normally

**Comment by jiang131072** [comment]
So is this a bug? or should we update the OpenCV exmple ?

**Reply by falkoschindler** [reply]
As far as we can tell, the OpenCV example is working as it is with and without `reload`. When using OpenCV video capturing in your own project, you have to take special care to clean up any resources on shutdown, as shown in `disconnect`, `handle_sigint`, and `cleanup`. If you don't do it correctly, setting `reload=False` might be a reasonable workaround.

**Reply by jiang131072** [reply]
I tried the  [Opencv webcam example](https://github.com/zauberzeug/nicegui/blob/bf4a0fd325975f088009846aad5a93f3786fc799/examples/opencv_webcam/main.py), but I'm getting these errors:

```log
[ WARN:0@0.525] global cap_v4l.cpp:999 open VIDEOIO(V4L2:/dev/video0): can't open camera by index
[ERROR:0@0.526] global obsensor_uvc_stream_channel.cpp:158 getStreamChannelGroup Camera index out of range
```

Setting `reload=False` worked.

I added `print(f"Video Capture 0 Opened: {video_capture.isOpened()}")` before `ui.run()`:

```log
Video Capture 0 Opened: True
[ WARN:0@0.562] global cap_v4l.cpp:999 open VIDEOIO(V4L2:/dev/video0): can't open camera by index
[ERROR:0@0.562] global obsensor_uvc_stream_channel.cpp:158 getStreamChannelGroup Camera index out of range
Video Capture 0 Opened: False
NiceGUI ready to go on http://localhost:8080, http://100.104.109.118:8080, and http://172.30.35.48:8080
```

After reloading, the same happened:

```log
WARNING:  WatchFiles detected changes in 'tmp/test.py'. Reloading...
[ WARN:0@0.576] global cap_v4l.cpp:999 open VIDEOIO(V4L2:/dev/video0): can't open camera by index
[ERROR:0@0.576] global obsensor_uvc_stream_channel.cpp:158 getStreamChannelGroup Camera index out of range
Video Capture 0 Opened: False
NiceGUI ready to go on http://localhost:8080, http://100.104.109.118:8080, and http://172.30.35.48:8080
```

It looks like the capture is mistakenly released.

**Reply by rodja** [reply]
The capture should be released on shutdown because it is created on startup. What is your operating system? Maybe releasing takes more time? Does it work if you put a `await asyncio.sleep(2)` after the `video_capture.release()` in line 64?

**Reply by jiang131072** [reply]
`await asyncio.sleep(2)` doesn't help. I am on Archlinux. `v4l2-ctl --list-devices` output:
```log
Integrated Camera: Integrated C (usb-0000:35:00.3-3):
        /dev/video0
        /dev/video1
        /dev/media0
```

**Reply by rodja** [reply]
Can anyone else reproduce the problem from @jiang131072 with Linux?

**Reply by rodja** [reply]
Another idea: Try replacing the top-level `video_capture` assignment with

```py
# OpenCV is used to access the webcam.
def setup() -> None:
    global video_capture
    video_capture = cv2.VideoCapture(0)

app.on_startup(setup)
```

That way, the `VideoCapture` device will only be allocated when the server really starts, and not once when the main reloader process starts and a second time when it starts the server.

**Reply by jiang131072** [reply]
That works.

**Reply by rodja** [reply]
Great. I updated the example in PR #3619 with a slightly improved variant of the `on_startup` technique to avoid the global variable. Does it work for you, too?

**Reply by jiang131072** [reply]
yes.
